pydra/utils/hash.py

```diff
@@ -52,19 +54,144 @@
+                appname="pydra",
+                appauthor="nipype",
+                version=__version__,
+            )
+        return location
+
+    # the default needs to be an instance method
+    @location.default
+    def _location_default(self):
+        return self.location_default()
+
+    @location.validator
+    def location_validator(self, _, location):
+        if not os.path.isdir(location):
+            raise ValueError(
+                f"Persistent cache location '{location}' is not a directory"
+            )
+
+    @cleanup_period.default
+    def cleanup_period_default(self):
+        return int(os.environ.get(self.CLEANUP_ENV_VAR, 30))
+
+    def get_or_calculate_hash(self, key: CacheKey, calculate_hash: ty.Callable) -> Hash:
+        """Check whether key is present in the persistent cache store and return it if so.
+        Otherwise use `calculate_hash` to generate the hash and save it in the persistent
+        store.
+
+        Parameters
+        ----------
+        key : CacheKey
+            locally unique key (e.g. to the host) used to lookup the corresponding hash
+            in the persistent store
+        calculate_hash : ty.Callable
+            function to calculate the hash if it isn't present in the persistent store
+
+        Returns
+        -------
+        Hash
+            _description_
+        """
+        try:
+            return self._hashes[key]
+        except KeyError:
+            pass
+        key_path = self.location / blake2s(str(key).encode()).hexdigest()
+        with SoftFileLock(key_path.with_suffix(".lock")):
+            if key_path.exists():
+                return Hash(key_path.read_bytes())
+            hsh = calculate_hash()
+            key_path.write_bytes(hsh)
+        return Hash(hsh)
+
+    def clean_up(self):
+        """Cleans up old hash caches that haven't been accessed in the last 30 days"""
+        now = datetime.now()
+        for path in self.location.iterdir():
+            if path.name.endswith(".lock"):
+                continue
+            days = (now - datetime.fromtimestamp(path.lstat().st_atime)).days
+            if days > self.cleanup_period:
+                path.unlink()
+
+    @classmethod
+    def from_path(
+        cls, path: ty.Union[Path, str, "PersistentCache", None]
+    ) -> "PersistentCache":
+        if isinstance(path, PersistentCache):
+            return path
+        return PersistentCache(path)
+
+
+@attrs.define
+class Cache:
+    persistent: ty.Optional[PersistentCache] = attrs.field(
+        default=None,
+        converter=PersistentCache.from_path,  # type: ignore[misc]
+    )
+    _hashes: ty.Dict[int, Hash] = attrs.field(factory=dict)
+
+    def __getitem__(self, object_id: int) -> Hash:
+        return self._hashes[object_id]
+
+    def __setitem__(self, object_id: int, hsh: Hash):
+        self._hashes[object_id] = hsh
+
+    def __contains__(self, object_id):
+        return object_id in self._hashes
```

Comment:
`self.persistent` never gets used. How is this supposed to work?

Reply:
It is used on L241 (https://github.com/tclose/pydra/blob/a031ea5dfc1571d2171e0df4157bb97114b9b1eb/pydra/utils/hash.py#L241)